libyul/optimiser/README.md

```diff
@@ -82,72 +175,405 @@ function calls can only appear in the right-hand side of a variable declaration,
+before. If the code is in pseudo-SSA form,
+the values of variables are available for a longer time and thus we
+have a higher chance of expressions to be replaceable.
+
+The expression simplifier will be able to perform better replacements
+if the common subexpression eliminator was run right before it.
+
+### Expression Simplifier
+
+The Expression Simplifier uses the Dataflow Analyzer and makes use
+of a list of equivalence transforms on expressions like ``X + 0 -> X``
+to simplify the code.
+
+It tries to match patterns like ``X + 0`` on each subexpression.
+During the matching procedure, it resolves variables to their currently
+assigned expressions to be able to match more deeply nested patters
+even when the code is in pseudo-SSA form.
+
+Some of the patterns like ``X - X -> 0`` can only be applied as long
+as ``X`` is movable, because it would remove its potential side-effects.
+Since variable references are always movable, even if their current
+value might not be, the Expression Simplifier is again more powerful
+in split or pseudo-SSA form.
+
+## Statement-Scale Simplifications
+
+### Unused Pruner
+
+This step removes the definitions of all functions that are never referenced.
+
+It also removes the declaration of variables that are never referenced.
+If the declaration assigns a value that is not movable, the expression is retained,
+but its value is discarded.
+
+All movable expression statements are removed.
+
+### Structural Simplifier
+
+This is a general step that performs various kinds of simplifications on
+a structural level:
+
+ - replace if statement with empty body by ``pop(condition)``
+ - replace if statement with true condition by its body
+ - remove if statement with false condition
+ - turn switch with single case into if
+ - replace switch with only default case by ``pop(expression)`` and body
+ - replace switch with literal expression by matching case body
+ - replace for loop with false condition by its initialization part
+
+This component uses the Dataflow Analyzer.
+
+### Equivalent Function Combiner
+
+If two functions are syntactically equivalent, while allowing variable
+renaming but not any re-ordering, then any reference to one of the
+functions is replaced by the other.
+
+The actual removal of the function is performed by the Unused Pruner.
+
+### Block Flattener
+
+This stage eliminates nested blocks by inserting the statement in the
+inner block at the appropriate place in the outer block:
+
+    {
+        let x := 2
+        {
+            let y := 3
+            mstore(x, y)
+        }
+    }
+
+is transformed to
+
+    {
+        let x := 2
+        let y := 3
+        mstore(x, y)
+    }
+
+As long as the code is disambiguated, this does not cause a problem because
+the scopes of variables can only grow.
+
+## Function Inlining
+
+### Functional Inliner
+
+The functional inliner performs function inlining such that the result of the inlining is a
```

Comment:
```suggestion
The result of function inlining is a
```

Reply:
I think this would make it too simple. This sentence differentiates the functional inliner from the full inliner. I'll try to find a better wording...